File: code/demo_analysis.R
#!/usr/bin/env Rscript

# Setup paths
args <- commandArgs(trailingOnly = FALSE)
script_path <- sub("^--file=", "", args[grep("^--file=", args)])
script_dir <- dirname(normalizePath(script_path))
proj_dir <- dirname(script_dir)
out_dir <- file.path(proj_dir, "code_out", "demo_analysis")
dir.create(out_dir, recursive = TRUE, showWarnings = FALSE)
setwd(out_dir)

# ---------------------------------------------------------------------
# Basic ggplot2
# ---------------------------------------------------------------------

# Load packages
library(tidyverse)

# Generate sample data
set.seed(42)
sample_data <- tibble(
    sample_id = paste0("Sample_", 1:20),
    condition = rep(c("Control", "Treatment"), each = 10),
    expression = c(rnorm(10, mean = 5, sd = 1.5), 
                   rnorm(10, mean = 8, sd = 1.5))
)

# Save data
write_csv(sample_data, "sample_data.csv")

# Create and save plot
p <- ggplot(sample_data, aes(x = condition, y = expression, fill = condition)) +
    geom_boxplot() +
    geom_jitter(width = 0.2, alpha = 0.5) +
    theme_minimal() +
    labs(title = "Gene Expression: Control vs Treatment",
         x = "Condition", y = "Expression Level") +
    scale_fill_brewer(palette = "Set2")

ggsave("expression_boxplot.png", p, width = 8, height = 6, dpi = 300)

# ---------------------------------------------------------------------
# Seurat analysis section
# ---------------------------------------------------------------------

# Configure reticulate to use conda Python (for Leiden algorithm)
library(reticulate)
reticulate_python <- Sys.getenv("RETICULATE_PYTHON")
if (reticulate_python != "") {
    use_python(reticulate_python, required = TRUE)
}

library(Seurat)

# Load bundled pbmc_small dataset
data("pbmc_small")

# Update to Seurat v5 format and convert assay to v5
pbmc_small <- UpdateSeuratObject(pbmc_small)
pbmc_small[["RNA"]] <- as(pbmc_small[["RNA"]], Class = "Assay5")

pbmc_small <- NormalizeData(pbmc_small)
pbmc_small <- FindVariableFeatures(pbmc_small)
pbmc_small <- ScaleData(pbmc_small)
pbmc_small <- RunPCA(pbmc_small)
pbmc_small <- FindNeighbors(pbmc_small, dims = 1:10)
# Clustering with Leiden algorithm
pbmc_small <- FindClusters(pbmc_small, algorithm = 4, resolution = 0.5)
pbmc_small <- RunUMAP(pbmc_small, dims = 1:10)

# Create UMAP plot and save as PDF
pdf("umap_leiden_clusters.pdf", width = 8, height = 6)
p_umap <- DimPlot(pbmc_small, reduction = "umap", label = TRUE) + 
    ggplot2::labs(title = "UMAP of Leiden Clusters")
print(p_umap)
dev.off()

# Summary of clusters
cat("Clustering complete. Cluster distribution:\n")
print(table(Idents(pbmc_small)))

cat("Done! Results in:", out_dir, "\n")

File: code/demo_analysis.py
#!/usr/bin/env python3
"""Simple data analysis demo using pandas and numpy"""

from pathlib import Path
import pandas as pd
import numpy as np

# Setup paths
script_dir = Path(__file__).resolve().parent
out_dir = script_dir.parent / "code_out" / "demo_analysis"
out_dir.mkdir(parents=True, exist_ok=True)

# Generate sample experimental data
np.random.seed(42)
n_samples = 20

data = {
    'sample_id': [f'Sample_{i:02d}' for i in range(1, n_samples + 1)],
    'group': ['Control'] * 10 + ['Treatment'] * 10,
    'measurement_1': np.random.normal(100, 15, n_samples),
    'measurement_2': np.random.normal(50, 10, n_samples)
}

# Create DataFrame
df = pd.DataFrame(data)

# Calculate summary statistics
df['mean_measurement'] = df[['measurement_1', 'measurement_2']].mean(axis=1)
df['ratio'] = df['measurement_1'] / df['measurement_2']

# Group statistics
summary = df.groupby('group').agg({
    'measurement_1': ['mean', 'std'],
    'measurement_2': ['mean', 'std'],
    'mean_measurement': ['mean', 'std']
}).round(2)

# Save results
df.to_csv(out_dir / "sample_measurements.csv", index=False)
summary.to_csv(out_dir / "group_summary.csv")

print(f"Done! Results in: {out_dir}")
print("\nSample data (first 5 rows):")
print(df.head().to_string(index=False))
print("\nGroup summary statistics:")
print(summary)
